lib/trie/trie.go

```diff
@@ -167,675 +177,888 @@ func (t *Trie) Entries() map[string][]byte {
 	return entries(t.root, nil, make(map[string][]byte))
 }
 
-func entries(current Node, prefix []byte, kv map[string][]byte) map[string][]byte {
-	switch c := current.(type) {
-	case *node.Branch:
-		if c.Value != nil {
-			kv[string(codec.NibblesToKeyLE(append(prefix, c.Key...)))] = c.Value
-		}
-		for i, child := range c.Children {
-			entries(child, append(prefix, append(c.Key, byte(i))...), kv)
-		}
-	case *node.Leaf:
-		kv[string(codec.NibblesToKeyLE(append(prefix, c.Key...)))] = c.Value
+func entries(parent Node, prefix []byte, kv map[string][]byte) map[string][]byte {
+	if parent == nil {
+		return kv
+	}
+
+	if parent.Type() == node.LeafType {
+		parentKey := parent.GetKey()
+		fullKeyNibbles := make([]byte, 0, len(prefix)+len(parentKey))
+		fullKeyNibbles = append(fullKeyNibbles, prefix...)
+		fullKeyNibbles = append(fullKeyNibbles, parentKey...)
+		keyLE := string(codec.NibblesToKeyLE(fullKeyNibbles))
+		kv[keyLE] = parent.GetValue()
 		return kv
 	}
 
+	// Branch with/without value
+	branch := parent.(*node.Branch)
+
+	if branch.Value != nil {
+		fullKeyNibbles := make([]byte, 0, len(prefix)+len(branch.Key))
+		fullKeyNibbles = append(fullKeyNibbles, prefix...)
+		fullKeyNibbles = append(fullKeyNibbles, branch.Key...)
+		keyLE := string(codec.NibblesToKeyLE(fullKeyNibbles))
+		kv[keyLE] = branch.Value
+	}
+
+	for i, child := range branch.Children {
+		childPrefix := make([]byte, 0, len(prefix)+len(branch.Key)+1)
+		childPrefix = append(childPrefix, prefix...)
+		childPrefix = append(childPrefix, branch.Key...)
+		childPrefix = append(childPrefix, byte(i))
+		entries(child, childPrefix, kv)
+	}
+
 	return kv
 }
 
-// NextKey returns the next key in the trie in lexicographic order. It returns nil if there is no next key
-func (t *Trie) NextKey(key []byte) []byte {
-	k := codec.KeyLEToNibbles(key)
+// NextKey returns the next key in the trie in lexicographic order.
+// It returns nil if no next key is found.
+func (t *Trie) NextKey(keyLE []byte) (nextKeyLE []byte) {
+	prefix := []byte(nil)
+	key := codec.KeyLEToNibbles(keyLE)
 
-	next := nextKey(t.root, nil, k)
-	if next == nil {
+	nextKey := findNextKey(t.root, prefix, key)
+	if nextKey == nil {
 		return nil
 	}
 
-	return codec.NibblesToKeyLE(next)
+	nextKeyLE = codec.NibblesToKeyLE(nextKey)
+	return nextKeyLE
 }
 
-func nextKey(curr Node, prefix, key []byte) []byte {
-	switch c := curr.(type) {
-	case *node.Branch:
-		fullKey := append(prefix, c.Key...)
-		var cmp int
-		if len(key) < len(fullKey) {
-			if bytes.Compare(key, fullKey[:len(key)]) == 1 { // arg key is greater than full, return nil
-				return nil
-			}
+func findNextKey(parent Node, prefix, searchKey []byte) (nextKey []byte) {
```

Comment:
I prefer the way it was previously. I find switch case more readable. I also prefer unnamed returns. Also why change the function name?